merlin-server/sql/merlin/hasura_functions.sql

```diff
@@ -231,3 +231,91 @@ begin
     select * from hasura_functions.get_conflicting_activities(_merge_request_id);
 end
 $$;
+
+-- Hasura functions for handling anchors during delete
+create table hasura_functions.delete_anchor_return_value(
+  affected_row activity_directive,
+  change_type text
+);
+create function hasura_functions.delete_activity_by_pk_reanchor_plan_start(_activity_id int, _plan_id int)
+  returns setof hasura_functions.delete_anchor_return_value
+  strict
+language plpgsql as $$
+  begin
+    if not exists(select id from public.activity_directive where (id, plan_id) = (_activity_id, _plan_id)) then
+      raise exception 'Activity Directive % does not exist in Plan %', _activity_id, _plan_id;
+    end if;
+
+    return query
+      with updated as (
+        select public.anchor_direct_descendents_to_plan(_activity_id := _activity_id, _plan_id := _plan_id)
+      )
+      select updated.*, 'updated'
+        from updated;
+
+    return query
+      with deleted as (
+        delete from activity_directive where (id, plan_id) = (_activity_id, _plan_id) returning *
+      )
+      select (deleted.id, deleted.plan_id, deleted.name, deleted.tags, deleted.source_scheduling_goal_id,
+              deleted.created_at, deleted.last_modified_at, deleted.start_offset, deleted.type, deleted.arguments,
+              deleted.last_modified_arguments_at, deleted.metadata, deleted.anchor_id, deleted.anchored_to_start)::activity_directive, 'deleted' from deleted;
+  end
+$$;
+
+create function hasura_functions.delete_activity_by_pk_reanchor_to_anchor(_activity_id int, _plan_id int)
+  returns setof hasura_functions.delete_anchor_return_value
+  strict
+  language plpgsql as $$
+begin
+  if not exists(select id from public.activity_directive where (id, plan_id) = (_activity_id, _plan_id)) then
+    raise exception 'Activity Directive % does not exist in Plan %', _activity_id, _plan_id;
+  end if;
+
+    return query
+      with updated as (
+        select public.anchor_direct_descendents_to_ascendant_anchor(_activity_id := _activity_id, _plan_id := _plan_id)
+      )
+      select updated.*, 'updated'
+        from updated;
+    return query
```

Comment:
I did not know that pl/pgsql allows multiple return statements in a function. Nifty!

Reply:
Yeah, because we're returning a `setof`, `return query` and `return next` contribute to the results set without actually terminating the function. [Here's the docs on it](https://www.postgresql.org/docs/14/plpgsql-control-structures.html#PLPGSQL-STATEMENTS-RETURNING) if you're curious.